hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -100,6 +141,338 @@ func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hy
 	// Drain.
 	// Create Namespace/RBAC/ConfigMap/Pod in guest cluster.
 	// Mark Node as Upgrading.
+
+	// This check comes after the completion, so if no updates are in progress, if a node is degraded for
+	// whatever reason, we will not know until the next update, at which point hopefully the MCD is able
+	// to reconcile
+	for _, node := range nodes {
+		if node.Annotations[MachineConfigDaemonStateAnnotationKey] == MachineConfigDaemonStateDegraded {
+			setStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolCondition{
+				Type:               hyperv1.NodePoolUpdatingVersionConditionType,
+				Status:             corev1.ConditionFalse,
+				Reason:             hyperv1.NodePoolInplaceUpgradeFailedConditionReason,
+				Message:            fmt.Sprintf("Node %s in nodepool degraded: %v", node.Name, node.Annotations[MachineConfigDaemonReasonAnnotationKey]),
+				ObservedGeneration: nodePool.Generation,
+			})
+			return fmt.Errorf("degraded node found, cannot progress in-place upgrade")
+		}
+	}
+
+	// Create necessary update manifests, if they do not exist
+	err = r.createInPlaceUpdateManifests(ctx, hostedClusterClient, targetConfigVersionHash, ignEndpoint, caCertBytes, tokenBytes, nodePool)
+	if err != nil {
+		return fmt.Errorf("failed to create update manifests in hosted cluster: %v", err)
+	}
+
+	// Check the nodes to see if any need our help to progress drain
+	// TODO actually implement drain logic, likely as separate goroutines to monitor success
+	// TODO consider what happens if the desiredConfig has changed since the node last updated
+	for _, node := range nodes {
+		if node.Annotations[DesiredDrainerAnnotationKey] != node.Annotations[LastAppliedDrainerAnnotationKey] {
+			err = r.handleNodeDrainRequest(ctx, hostedClusterClient, node, node.Annotations[DesiredDrainerAnnotationKey])
+			if err != nil {
+				return fmt.Errorf("failed to create update manifests in hosted cluster: %v", err)
+			}
+			// TODO in the future, consider exiting here and let future syncs handle post-drain functions
+		}
+	}
+
+	// Find nodes that can be upgraded
+	// TODO add logic to honor maxUnavailable
+	nodesToUpgrade := getAllCandidateMachines(nodes, targetConfigVersionHash, 1)
+	err = r.setUpdateForNodes(ctx, hostedClusterClient, nodePool, nodesToUpgrade, targetConfigVersionHash)
+	if err != nil {
+		return fmt.Errorf("failed to set hosted nodes for inplace upgrade: %v", err)
+	}
+
+	return nil
+}
+
+func (r *NodePoolReconciler) setUpdateForNodes(ctx context.Context, hostedClusterClient client.Client, nodePool *hyperv1.NodePool, nodes []*corev1.Node, targetConfigVersionHash string) error {
```

Comment:
can we rename this to `performNodesUpgrade` so then we got  getNodesToUpgrade and performNodesUpgrade which is consistent